File: C++/password_generator/password_generator.cpp
#include <iostream>
#include <string>
#include <vector>
#include <random>
#include <algorithm>
#include <cctype>
#include <iomanip>

class PasswordGenerator {
private:
    std::random_device rd;
    std::mt19937 gen;
    
    // Character sets
    const std::string lowercase = "abcdefghijklmnopqrstuvwxyz";
    const std::string uppercase = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    const std::string digits = "0123456789";
    const std::string symbols = "!@#$%^&*()_+-=[]{}|;:,.<>?";
    const std::string ambiguous = "0O1lI"; // Characters that can be confused
    
public:
    PasswordGenerator() : gen(rd()) {}
    
    struct PasswordConfig {
        int length = 12;
        bool includeLowercase = true;
        bool includeUppercase = true;
        bool includeDigits = true;
        bool includeSymbols = false;
        bool excludeAmbiguous = false;
        int minNumbers = 1;
        int minSymbols = 0;
        bool noRepeatingChars = false;
    };
    
    std::string generatePassword(const PasswordConfig& config) {
        std::string charset = buildCharset(config);
        
        if (charset.empty()) {
            throw std::runtime_error("No character types selected!");
        }
        
        std::string password;
        std::uniform_int_distribution<> dis(0, charset.length() - 1);
        
        // Ensure minimum requirements are met
        password += ensureMinimumRequirements(config);
        
        // Fill remaining length
        while (password.length() < config.length) {
            char nextChar = charset[dis(gen)];
            
            if (config.noRepeatingChars && 
                std::find(password.begin(), password.end(), nextChar) != password.end()) {
                continue;
            }
            
            password += nextChar;
        }
        
        // Shuffle the password to randomize position of required characters
        std::shuffle(password.begin(), password.end(), gen);
        
        return password;
    }
    
    std::vector<std::string> generateMultiple(const PasswordConfig& config, int count) {
        std::vector<std::string> passwords;
        passwords.reserve(count);
        
        for (int i = 0; i < count; ++i) {
            passwords.push_back(generatePassword(config));
        }
        
        return passwords;
    }
    
    double calculateStrength(const std::string& password) {
        double score = 0;
        int charsetSize = 0;
        
        bool hasLower = false, hasUpper = false, hasDigit = false, hasSymbol = false;
        
        for (char c : password) {
            if (std::islower(c)) hasLower = true;
            else if (std::isupper(c)) hasUpper = true;
            else if (std::isdigit(c)) hasDigit = true;
            else hasSymbol = true;
        }
        
        if (hasLower) charsetSize += 26;
        if (hasUpper) charsetSize += 26;
        if (hasDigit) charsetSize += 10;
        if (hasSymbol) charsetSize += 32;
        
        // Calculate entropy: log2(charsetSize^length)
        double entropy = password.length() * std::log2(charsetSize);
        
        // Convert to strength score (0-100)
        score = std::min(100.0, (entropy / 128.0) * 100.0);
        
        return score;
    }
    
    std::string getStrengthLabel(double strength) {
        if (strength < 25) return "Very Weak";
        if (strength < 50) return "Weak";
        if (strength < 75) return "Good";
        if (strength < 90) return "Strong";
        return "Very Strong";
    }

private:
    std::string buildCharset(const PasswordConfig& config) {
        std::string charset;
        
        if (config.includeLowercase) charset += lowercase;
        if (config.includeUppercase) charset += uppercase;
        if (config.includeDigits) charset += digits;
        if (config.includeSymbols) charset += symbols;
        
        if (config.excludeAmbiguous) {
            for (char c : ambiguous) {
                charset.erase(std::remove(charset.begin(), charset.end(), c), charset.end());
            }
        }
        
        return charset;
    }
    
    std::string ensureMinimumRequirements(const PasswordConfig& config) {
        std::string required;
        std::uniform_int_distribution<> lowerDis(0, lowercase.length() - 1);
        std::uniform_int_distribution<> upperDis(0, uppercase.length() - 1);
        std::uniform_int_distribution<> digitDis(0, digits.length() - 1);
        std::uniform_int_distribution<> symbolDis(0, symbols.length() - 1);
        
        // Add required numbers
        for (int i = 0; i < config.minNumbers && config.includeDigits; ++i) {
            required += digits[digitDis(gen)];
        }
        
        // Add required symbols
        for (int i = 0; i < config.minSymbols && config.includeSymbols; ++i) {
            required += symbols[symbolDis(gen)];
        }
        
        return required;
    }
};

void displayMenu() {
    std::cout << "\n========================================\n";
    std::cout << "       PASSWORD GENERATOR TOOL\n";
    std::cout << "========================================\n";
    std::cout << "1. Generate Single Password\n";
    std::cout << "2. Generate Multiple Passwords\n";
    std::cout << "3. Check Password Strength\n";
    std::cout << "4. Custom Password Configuration\n";
    std::cout << "5. Exit\n";
    std::cout << "========================================\n";
    std::cout << "Choose an option (1-5): ";
}

void displayPasswordWithStrength(const std::string& password, PasswordGenerator& generator) {
    double strength = generator.calculateStrength(password);
    std::string strengthLabel = generator.getStrengthLabel(strength);
    
    std::cout << "\nGenerated Password: " << password << "\n";
    std::cout << "Length: " << password.length() << " characters\n";
    std::cout << "Strength: " << std::fixed << std::setprecision(1) 
              << strength << "% (" << strengthLabel << ")\n";
    
    // Visual strength bar
    std::cout << "Strength Bar: [";
    int bars = static_cast<int>(strength / 5);
    for (int i = 0; i < 20; ++i) {
        if (i < bars) std::cout << "=";
        else std::cout << " ";
    }
    std::cout << "]\n";
}

int main() {
    PasswordGenerator generator;
    int choice;
    
    std::cout << "Welcome to the Advanced Password Generator!\n";
    
    while (true) {
        displayMenu();
        std::cin >> choice;
        
        switch (choice) {
            case 1: {
                // Generate single password with default settings
                PasswordGenerator::PasswordConfig config;
                std::string password = generator.generatePassword(config);
                displayPasswordWithStrength(password, generator);
                break;
            }
            
            case 2: {
                // Generate multiple passwords
                int count;
                std::cout << "How many passwords to generate? ";
                std::cin >> count;
                
                PasswordGenerator::PasswordConfig config;
                auto passwords = generator.generateMultiple(config, count);
                
                std::cout << "\nGenerated " << count << " passwords:\n";
                std::cout << "--------------------------------\n";
                for (size_t i = 0; i < passwords.size(); ++i) {
                    std::cout << (i + 1) << ". " << passwords[i] << "\n";
                }
                break;
            }
            
            case 3: {
                // Check password strength
                std::string password;
                std::cout << "Enter password to check: ";
                std::cin >> password;
                
                double strength = generator.calculateStrength(password);
                std::string strengthLabel = generator.getStrengthLabel(strength);
                
                std::cout << "\nPassword Analysis:\n";
                std::cout << "Password: " << password << "\n";
                std::cout << "Length: " << password.length() << " characters\n";
                std::cout << "Strength: " << std::fixed << std::setprecision(1) 
                          << strength << "% (" << strengthLabel << ")\n";
                break;
            }
            
            case 4: {
                // Custom configuration
                PasswordGenerator::PasswordConfig config;
                
                std::cout << "\n--- Custom Password Configuration ---\n";
                std::cout << "Password length (4-128): ";
                std::cin >> config.length;
                config.length = std::max(4, std::min(128, config.length));
                
                char choice;
                std::cout << "Include lowercase letters? (y/n): ";
                std::cin >> choice;
                config.includeLowercase = (choice == 'y' || choice == 'Y');
                
                std::cout << "Include uppercase letters? (y/n): ";
                std::cin >> choice;
                config.includeUppercase = (choice == 'y' || choice == 'Y');
                
                std::cout << "Include digits? (y/n): ";
                std::cin >> choice;
                config.includeDigits = (choice == 'y' || choice == 'Y');
                
                std::cout << "Include symbols? (y/n): ";
                std::cin >> choice;
                config.includeSymbols = (choice == 'y' || choice == 'Y');
                
                std::cout << "Exclude ambiguous characters (0,O,1,l,I)? (y/n): ";
                std::cin >> choice;
                config.excludeAmbiguous = (choice == 'y' || choice == 'Y');
                
                try {
                    std::string password = generator.generatePassword(config);
                    displayPasswordWithStrength(password, generator);
                } catch (const std::exception& e) {
                    std::cout << "Error: " << e.what() << "\n";
                }
                break;
            }
            
            case 5:
                std::cout << "Thank you for using Password Generator!\n";
                return 0;
                
            default:
                std::cout << "Invalid choice! Please try again.\n";
                break;
        }
        
        std::cout << "\nPress Enter to continue...";
        std::cin.ignore();
        std::cin.get();
    }
    
    return 0;
}

File: Rust/password_generator.rs
use std::io;
use rand::Rng;

struct PasswordConfig {
    length: usize,
    include_lowercase: bool,
    include_uppercase: bool,
    include_numbers: bool,
    include_symbols: bool,
    exclude_ambiguous: bool,
}

impl Default for PasswordConfig {
    fn default() -> Self {
        PasswordConfig {
            length: 12,
            include_lowercase: true,
            include_uppercase: true,
            include_numbers: true,
            include_symbols: false,
            exclude_ambiguous: false,
        }
    }
}

struct PasswordGenerator {
    lowercase: String,
    uppercase: String,
    numbers: String,
    symbols: String,
    ambiguous: String,
}

impl PasswordGenerator {
    fn new() -> Self {
        PasswordGenerator {
            lowercase: "abcdefghijklmnopqrstuvwxyz".to_string(),
            uppercase: "ABCDEFGHIJKLMNOPQRSTUVWXYZ".to_string(),
            numbers: "0123456789".to_string(),
            symbols: "!@#$%^&*()_+-=[]{}|;:,.<>?".to_string(),
            ambiguous: "0O1lI".to_string(),
        }
    }

    fn build_charset(&self, config: &PasswordConfig) -> String {
        let mut charset = String::new();

        if config.include_lowercase {
            charset.push_str(&self.lowercase);
        }
        if config.include_uppercase {
            charset.push_str(&self.uppercase);
        }
        if config.include_numbers {
            charset.push_str(&self.numbers);
        }
        if config.include_symbols {
            charset.push_str(&self.symbols);
        }

        if config.exclude_ambiguous {
            charset = charset
                .chars()
                .filter(|c| !self.ambiguous.contains(*c))
                .collect();
        }

        charset
    }

    fn generate_password(&self, config: &PasswordConfig) -> Result<String, String> {
        let charset = self.build_charset(config);
        
        if charset.is_empty() {
            return Err("No character types selected!".to_string());
        }

        let mut rng = rand::thread_rng();
        let charset_chars: Vec<char> = charset.chars().collect();
        
        let password: String = (0..config.length)
            .map(|_| {
                let idx = rng.gen_range(0..charset_chars.len());
                charset_chars[idx]
            })
            .collect();

        Ok(password)
    }

    fn calculate_strength(&self, password: &str) -> f64 {
        let mut charset_size = 0;
        let mut has_lower = false;
        let mut has_upper = false;
        let mut has_digit = false;
        let mut has_symbol = false;

        for ch in password.chars() {
            if ch.is_ascii_lowercase() && !has_lower {
                has_lower = true;
                charset_size += 26;
            } else if ch.is_ascii_uppercase() && !has_upper {
                has_upper = true;
                charset_size += 26;
            } else if ch.is_ascii_digit() && !has_digit {
                has_digit = true;
                charset_size += 10;
            } else if !ch.is_alphanumeric() && !has_symbol {
                has_symbol = true;
                charset_size += 32;
            }
        }

        if charset_size == 0 {
            return 0.0;
        }

        // Calculate entropy: log2(charset_size^length)
        let entropy = password.len() as f64 * (charset_size as f64).log2();
        
        // Convert to strength score (0-100)
        (entropy / 128.0 * 100.0).min(100.0)
    }

    fn get_strength_label(&self, strength: f64) -> &str {
        match strength {
            s if s < 25.0 => "Very Weak",
            s if s < 50.0 => "Weak",
            s if s < 75.0 => "Good",
            s if s < 90.0 => "Strong",
            _ => "Very Strong",
        }
    }

    fn display_strength_bar(&self, strength: f64) {
        let bars = (strength / 5.0) as usize;
        print!("Strength Bar: [");
        for i in 0..20 {
            if i < bars {
                print!("=");
            } else {
                print!(" ");
            }
        }
        println!("]");
    }
}

fn get_user_config() -> PasswordConfig {
    let mut config = PasswordConfig::default();

    println!("\n=== Password Configuration ===");
    
    print!("Password length (4-128) [default: 12]: ");
    io::Write::flush(&mut io::stdout()).unwrap();
    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();
    if let Ok(length) = input.trim().parse::<usize>() {
        if (4..=128).contains(&length) {
            config.length = length;
        }
    }

    print!("Include lowercase letters? (y/n) [default: y]: ");
    io::Write::flush(&mut io::stdout()).unwrap();
    input.clear();
    io::stdin().read_line(&mut input).unwrap();
    config.include_lowercase = !input.trim().eq_ignore_ascii_case("n");

    print!("Include uppercase letters? (y/n) [default: y]: ");
    io::Write::flush(&mut io::stdout()).unwrap();
    input.clear();
    io::stdin().read_line(&mut input).unwrap();
    config.include_uppercase = !input.trim().eq_ignore_ascii_case("n");

    print!("Include numbers? (y/n) [default: y]: ");
    io::Write::flush(&mut io::stdout()).unwrap();
    input.clear();
    io::stdin().read_line(&mut input).unwrap();
    config.include_numbers = !input.trim().eq_ignore_ascii_case("n");

    print!("Include symbols? (y/n) [default: n]: ");
    io::Write::flush(&mut io::stdout()).unwrap();
    input.clear();
    io::stdin().read_line(&mut input).unwrap();
    config.include_symbols = input.trim().eq_ignore_ascii_case("y");

    print!("Exclude ambiguous characters (0,O,1,l,I)? (y/n) [default: n]: ");
    io::Write::flush(&mut io::stdout()).unwrap();
    input.clear();
    io::stdin().read_line(&mut input).unwrap();
    config.exclude_ambiguous = input.trim().eq_ignore_ascii_case("y");

    config
}

fn main() {
    println!("🔐 Secure Password Generator");
    println!("============================");

    let generator = PasswordGenerator::new();

    loop {
        println!("\nOptions:");
        println!("1. Generate password with default settings");
        println!("2. Generate password with custom settings");
        println!("3. Check password strength");
        println!("4. Generate multiple passwords");
        println!("5. Exit");
        print!("Choose an option (1-5): ");
        io::Write::flush(&mut io::stdout()).unwrap();

        let mut input = String::new();
        io::stdin().read_line(&mut input).expect("Failed to read input");
        
        match input.trim() {
            "1" => {
                let config = PasswordConfig::default();
                match generator.generate_password(&config) {
                    Ok(password) => {
                        let strength = generator.calculate_strength(&password);
                        println!("\nGenerated Password: {}", password);
                        println!("Length: {} characters", password.len());
                        println!("Strength: {:.1}% ({})", strength, generator.get_strength_label(strength));
                        generator.display_strength_bar(strength);
                    }
                    Err(e) => println!("Error: {}", e),
                }
            }
            "2" => {
                let config = get_user_config();
                match generator.generate_password(&config) {
                    Ok(password) => {
                        let strength = generator.calculate_strength(&password);
                        println!("\nGenerated Password: {}", password);
                        println!("Length: {} characters", password.len());
                        println!("Strength: {:.1}% ({})", strength, generator.get_strength_label(strength));
                        generator.display_strength_bar(strength);
                    }
                    Err(e) => println!("Error: {}", e),
                }
            }
            "3" => {
                print!("Enter password to analyze: ");
                io::Write::flush(&mut io::stdout()).unwrap();
                let mut password = String::new();
                io::stdin().read_line(&mut password).unwrap();
                let password = password.trim();
                
                if !password.is_empty() {
                    let strength = generator.calculate_strength(password);
                    println!("\nPassword Analysis:");
                    println!("Password: {}", password);
                    println!("Length: {} characters", password.len());
                    println!("Strength: {:.1}% ({})", strength, generator.get_strength_label(strength));
                    generator.display_strength_bar(strength);
                }
            }
            "4" => {
                print!("How many passwords to generate? ");
                io::Write::flush(&mut io::stdout()).unwrap();
                let mut count_input = String::new();
                io::stdin().read_line(&mut count_input).unwrap();
                
                if let Ok(count) = count_input.trim().parse::<usize>() {
                    if count > 0 && count <= 20 {
                        let config = PasswordConfig::default();
                        println!("\nGenerated {} passwords:", count);
                        println!("------------------------");
                        
                        for i in 1..=count {
                            match generator.generate_password(&config) {
                                Ok(password) => println!("{}. {}", i, password),
                                Err(e) => println!("{}. Error: {}", i, e),
                            }
                        }
                    } else {
                        println!("Please enter a number between 1 and 20.");
                    }
                } else {
                    println!("Invalid number!");
                }
            }
            "5" => {
                println!("Thank you for using the Password Generator!");
                break;
            }
            _ => {
                println!("Invalid choice! Please enter 1-5.");
            }
        }
    }
}